### PR TITLE
feat: add iterator over compute paths for MerklePath

### PR DESCRIPTION
**Pull request by hackaugusto**
## Describe your changes

This adds an iterator to expose the intermediary values computed while verifying a Merkle path. This is useful to visualize the values during debugging, I introduced this while looking at https://github.com/0xPolygonMiden/crypto/issues/119#issuecomment-1495955729 .

It also adds some tests for the MerklePath, it seems we didn't have direct tests for it.

## Checklist before requesting a review
- Repo forked and branch created from `next` according to naming convention.
- Commit messages and codestyle follow [conventions](./CONTRIBUTING.md).
- Relevant issues are linked in the PR description.
- Tests added for new functionality.
- Documentation/comments updated according to changes.

### CODE REVIEW

File: src/merkle/path.rs [hunk 1]
```diff
@@ -24,13 +24,31 @@ impl MerklePath {
...
-            index.move_up();
-            Rpo256::merge(&input).into()
-        })
+        self.compute_parents(index_value, node)
+            .last()
+            .unwrap_or(node)
+    }
+
+    /// Compute parent nodes based on the provided data
+    pub fn compute_parents(&self, index_value: u64, node: Word) -> impl Iterator<Item = Word> + '_ {
```
Comment (vlopes11) [inline src/merkle/path.rs:33]: The internal implementation of the function might not be immediately obvious if we just check the function name + docs.

We might want to be more verbose there. A suggestion is:

"
Returns an iterator that will traverse from the leaf towards the root, yielding the parent of `node` for every iteration. 
"

File: src/merkle/path.rs [hunk 2]
```diff
@@ -24,13 +24,31 @@ impl MerklePath {
...
+        struct State {
+            index: NodeIndex,
+            curr_node: RpoDigest,
+        }
```
Comment (vlopes11) [inline src/merkle/path.rs:37]: Isn't overkill to add a struct here when this could easily be a tuple?

To give further context; I think it's always best to create wrapping structs when we have exposed tuples. But in this case, this is completely internal and the tuple won't ever leak

File: src/merkle/path.rs [hunk 3]
```diff
@@ -106,3 +124,74 @@ pub struct RootPath {
     /// The path from `value` to `root` (exclusive).
     pub path: MerklePath,
 }
+
+#[cfg(test)]
+mod test {
+    use crate::merkle::{int_to_node, MerklePath, Rpo256, Vec, Word};
```
Comment (vlopes11) [inline src/merkle/path.rs:130]: We might want to add some edge cases such as when we have an empty path or single node to assert it won't panic + return a consistent result.

We also changed the implementation of `compute_root`, so it would be good to add some tests cases for that as well

**Review by vlopes11 — COMMENTED** [review]
Looks good overall; a couple of optional nits

File: src/merkle/path.rs (hunk 2 above)
Comment (hackaugusto) [reply src/merkle/path.rs:37]: The type annotations are required for the `into` calls.

File: src/merkle/path.rs (hunk 1 above)
Comment (hackaugusto) [reply src/merkle/path.rs:33]: changed

File: src/merkle/path.rs (hunk 3 above)
Comment (hackaugusto) [reply src/merkle/path.rs:130]: added a test for empty path, `compute_root` are implicitly test via `verify`

File: src/merkle/path.rs (hunk 2 above)
Comment (hackaugusto) [reply src/merkle/path.rs:38]: Yeah, looking over these two alternatives there is no advantage to either. Going to keep the original one since it is already there.

**Comment by bobbinth** [comment]
Closing this as I've implemented `MerklePath::inner_nodes()` in #139.
